apiserver/facades/client/application/application.go

```diff
@@ -957,27 +963,42 @@ func (api *APIBase) setCharmWithAgentValidation(
 	params setCharmParams,
 	url string,
 ) error {
-	newCharm, err := api.backend.Charm(url)
+	// TODO (stickupkid): This should be done, in the application service.
+	// We'll want to start migrating the logic to the service layer.
+	newOrigin, err := convertCharmOrigin(params.CharmOrigin)
 	if err != nil {
 		return errors.Trace(err)
 	}
-	oneApplication := params.Application
-	currentCharm, _, err := oneApplication.Charm()
+
+	newCharmID, err := api.getCharmID(ctx, url)
 	if err != nil {
-		api.logger.Debugf("Unable to locate current charm: %v", err)
+		return errors.Trace(err)
 	}
-	newOrigin, err := convertCharmOrigin(params.CharmOrigin)
+
+	newCharm, err := api.getCharm(ctx, newCharmID)
 	if err != nil {
 		return errors.Trace(err)
 	}
+
 	if api.modelInfo.Type == model.CAAS {
+		// TODO (stickupkid): This should be done, in the application service
+		// layer.
+		currentCharmID, err := api.getCharmIDByApplicationName(ctx, params.AppName)
```

Comment:
It is indeed. See my bullet point list in the description about inefficiencies. I'm not going to change the logic right here and now though. This will be a future problem.